docs/admin/authentication.md

```diff
@@ -222,44 +222,103 @@ from the OAuth2 [token response](https://openid.net/specs/openid-connect-core-1_
-Or, you can run your own Identity Provider, such as CoreOS [dex](https://github.com/coreos/dex), [Keycloak](https://github.com/keycloak/keycloak) or CloudFoundry [UAA](https://github.com/cloudfoundry/uaa).
+Or, you can run your own Identity Provider, such as CoreOS [dex](https://github.com/coreos/dex), [Keycloak](https://github.com/keycloak/keycloak), CloudFoundry [UAA](https://github.com/cloudfoundry/uaa), or Tremolo Secutity's [OpenUnison](https://github.com/tremolosecurity/openunison).
+
+For an identity provider to work with Kubernetes it must:
+
+1.  Support [OpenID connect discovery](https://openid.net/specs/openid-connect-discovery-1_0.html); not all do.
+2.  Run in TLS with non-obsolete ciphers
+3.  Have a CA signed certificate (even if the CA is not a commercial CA or is self signed)
 
-The provider needs to support [OpenID connect discovery](https://openid.net/specs/openid-connect-discovery-1_0.html); not all do.
+A note about requirement #3 above, requiring a CA signed certificate.  If you deploy your own identity provider (as apposed to one of the cloud providers like Goolge or Microsoft) you MUST have your identity provider's web server certificate signed by a CA.  This is an issue with GoLang's TLS client implementation not being able to verify a self-signed certificate.  If you don't have a CA handy, you can use this script from the CoreOS team to create a simple CA and a signed certificate and key pair - https://github.com/coreos/dex/blob/1ee5920c54f5926d6468d2607c728b71cfe98092/examples/k8s/gencert.sh or this script based on it that will generate sha256 certs with a longer life and larger keysize https://raw.githubusercontent.com/TremoloSecurity/openunison-qs-kubernetes/master/makecerts.sh.
 
 Setup instructions for specific systems:
 
 - [UAA](http://apigee.com/about/blog/engineering/kubernetes-authentication-enterprise)
 - [Dex](https://speakerdeck.com/ericchiang/kubernetes-access-control-with-dex)
+- [OpenUnison](https://github.com/TremoloSecurity/openunison-qs-kubernetes)
+
+#### Using kubectl
+
+##### Option 1 - OIDC Authenticator
+
+The first option is to use the oidc authenticator.  This authenticator takes your `id_token`, `refresh_token` and your OIDC `client_secret` and will refresh your token automatically.  Once you have authenticated to your identity provider:
+
+```
+kubectl config set-credentials USER_NAME --auth-provider=oidc
+kubectl config set-credentials USER_NAME --auth-provider-arg=idp-issuer-url=( issuer url )
+kubectl config set-credentials USER_NAME --auth-provider-arg=client-id=( your client id )
+kubectl config set-credentials USER_NAME --auth-provider-arg=client-secret=( your client secret )
+kubectl config set-credentials USER_NAME --auth-provider-arg=refresh-token=( your refresh token )
+```
+Alternatively, you can add these configuration options manually to your `~/.kube/config`:
```

Comment:
@liggitt so you want me to show running the kubectl command that creates the config and the resulting config?